### ISSUE TITLE
Open large geotiff raster failed due to file pointer math overflow

### ISSUE THREAD

**Issue by TongZhai** [DotSpatial/DotSpatial#1334]
This issue is referenced in the issue #901. 
In the DemoMap application, when opening very large grid (2GB+), it failed 

![image](https://user-images.githubusercontent.com/16242714/79910073-08b7b880-83ec-11ea-8c48-ef46e57e0526.png)

In issue #901, @VectorZita suggested increasing pointer math variable during Seek operation to get over the overflow problem, I've tried this and it indeed got further, but there are other places also have this problem.

**Comment by TongZhai** [comment]
Also, opening large grid is very slow, in the above case, opening the 2GB+ grid took quite a while (a few minutes) to get to 50% loading and then it failed on the math pointer overflow exception. In contrast, QGIS loads this same large grid almost instantly.

**Comment by VectorZita** [comment]
I have used the large raster version of [this piece of data](https://www.naturalearthdata.com/downloads/10m-cross-blend-hypso/cross-blended-hypso-with-relief-water-drains-and-ocean-bottom/) to test a bit. Adding this to DotSpatial seems to be taking ~40 seconds using a 6-core ~2.6 GHz processor. Debug builds seem to have a large difference in performance from Release builds. Tried to place the file in an SSD or a classic HDD with no significant difference in performance. It seems that disk I/O is not a limiting factor. I will have to look into this a bit more.

Regardless, it seems pretty clear that the Pyramid creation is taking most of the time, especially the `DrawToBitmap` method of `DesktopRasterExt.cs`.

@jany-tenaj In the past (issue #940) you have mentioned a [commit of yours](https://github.com/DotSpatial/DotSpatial/commit/faa47c77f012df53fec8931550925322900f834e), which made an attempt to create pyramids using GDAL's built-in method. I see that this change has not been integrated into the latest version. Has this been tested, or what became of it in general? I assume GDAL would be more performant in building pyramids.

**Comment by VectorZita** [comment]
I have been looking into this for a while, intensely focusing on optimizing the `DrawToBitmap` and relevant routines. I have probably fought a bit too much with this, primarily because I have been thinking that there must be something obviously going wrong. Indeed... I think I have found the problem. I have been struggling to profile and optimize the `DrawToBitmap` method I mentioned, when it suddenly struck me...

If you create a very simple test project in c#.NET, create a byte array, about 10000 x 10000 and simply assign a value to all bytes, this already costs a **nontrivial amount of CPU time**. For me it's some half a second. With all blockwise reading/writing plus intermediate color calculations and stuff taking place within `DrawToBitmap` internally, this is taking **far far more**. The problem is not that it's taking that much time, of course. The problem is that **it should not happen in the first place**.

Let me explain myself a bit!
_______________________

Pyramids
====
Pyramids (overviews, or however you want to call them) are "down-sampled" copies of the same raster. A 20000 x 20000 pixel raster can be downsampled to 10000 x 10000, 5000 x 5000 etc. so that lower resolutions are used when necessary.

What data is down-sampled?
===
**Normally**, the raw raster has 1 or more _bands_ and each band has a value type (i.e. `byte`, `short`, etc.) and each cell/pixel has a **value**. A very specific value! This **raw data** is down-sampled using _some method_ (e.g. Nearest Neighbor) and this lower-resolution re-sampled data is written as the "Pyramid".

Where are we going wrong then?
=====
The current DotSpatial implementation of `RasterLayer` method `CreatePyramidImage` goes one assumption _too far_. It creates **rendered Pyramids**. Instead of writing raw data to file, data is **rendered** (using the current symbolizer) and the rendered data is down-sampled instead. Rendering itself takes time, **a lot of time**. In doubt? Open up QGIS, load a huge raster, change the symbology (e.g. use a simple hill-shade based on some band), then export the raster to GeoTIFF and _make sure to choose_ **Renderd Image** instead of **Raw Data** in the settings (should be at the top of the dialog box). Rendering is not a split-second task for large rasters. Apparently...

DotSpatial Symbolizer-rendering definitely takes some more time, being managed code and all, but the point is that **this** is the vastly time consuming task taking place while waiting for the image to load... Pyramids are actually **rendered** to file and **then** down-sampled. There is probably a **much more efficient method**, which I assume is what happens in QGIS and other GIS applications. Pyramids are most likely created to speed up down-sampled raster retrieval, but the retrieved part is rendered **afterwards**, on the fly.

This way, rendering is always carried out in real-time (of course this slows it down a bit), but only for the small extent that is being rendered, plus only from the down-sampled raster. This probably never leads to a necessity to render more than the maximum number of pixels hosted by the current screen resolution. This, effectively, means that we can get away with much faster pyramid-creation at the expense of somewhat slower real-time rendering. If real-time rendering is not too much affected, this should be hugely beneficial.

I am trying to experiment with the classes involved, in order to put this theory into practice. I will hopefully check back with more results soon!

**Comment by jany-tenaj** [comment]
@VectorZita This sounds as if you're trying to create the pyramids yourself instead of simply trying to get the gdal version running?

**Comment by VectorZita** [comment]
@jany-tenaj Not really, I am only trying to document what is _currently_ going on inside the core library and experiment with the various tools already available. I have stumbled upon various bugs, mostly minor, but which are enough to break various tools. I will try to document those soon as well.

I clearly prefer a GDAL-supported implementation, however most DotSpatial classes appear to have been created with an intent to work on their own as an alternative. For example `MapRasterLayer` can also create pyramids "by itself", in the absence of GDAL support. I don't know your thoughts on whether we should also attempt to provide an improved built-in managed implementation as an alternative. In that case, I assume building overviews _automagically_ when loading a MapRasterLayer should not create _internal overviews_, thus altering the file, rather probably external overviews. What do you think?

For now, I am trying to experiment with whatever is already available, including your old commit, in order to work out a simple demo workflow of loading a large image and getting it running.

**Comment by VectorZita** [comment]
OK, it seems that what I am thinking of doing (switching the raster rendering with a simple raw-data write and down-sampling) is already implemented in the `GdalImageProvider`. Therefore, there is what I have been missing all along.

```
GdalImageProvider gip = new GdalImageProvider();

var imageData = gip.Open(filename);

MapImageLayer imgLayer = new MapImageLayer(imageData);

map.Layers.Add(imgLayer);
```

Overviews are built externally as `**.mwi**` and `**.mwh**` files.

This should be faster for now. @TongZhai do give this a shot to see if the raster loads faster. You may have to adjust your `int` values to `long` once again because of the overflow problem, but thereafter, it should work out of the box. I am not entirely sure that the bug fixes I proposed earlier (#1326 and #1332) will also cover this case. If the raster does not appear immediately, you may need to zoom in a bit and let me know so that I can also look into this. However, I think these fixes should have this case covered as well.

I will try the older commit of @jany-tenaj now to see how this compares in terms of performance.

**Comment by VectorZita** [comment]
Yes. Using the `GdalImageProvider` of the [older commit](https://github.com/DotSpatial/DotSpatial/commit/faa47c77f012df53fec8931550925322900f834e) of @jany-tenaj is **far more performant**, GDAL is a lot faster indeed. @TongZhai do try to use this different version to compare for speed, using the simple code excerpt I posted above.

The only problem to pay attention to is that the unsuspecting user will see an increase in their file because GDAL overviews are built **internally**. At the moment, I do not have enough information on how to control this, therefore, @TongZhai, be warned, when testing for this, you may want to use a copy.

I assume GDAL will have accounted for the case of very large files... maybe overviews for very large files are built externally, but I will have to double-check that.

**Comment by VectorZita** [comment]
I made a more **fair** comparison between GDAL overview building and DotSpatial pyramids... DotSpatial spends more time because the file itself is re-written to the external pyramid file (i.e. scale zero, the large resolution raster itself). Ignoring the time taken to re-write the zero scale, the GDAL implementation is not **that much faster**, although the difference is still comparable (slightly less than 6 seconds for GDAL compared to slightly over 8 seconds for DotSpatial pyramids for the [large version of this piece of data](https://www.naturalearthdata.com/downloads/10m-cross-blend-hypso/cross-blended-hypso-with-relief-water-drains-and-ocean-bottom/)).

In terms of this, maybe writing the zero-scale could be avoided by simply creating a blind copy of the original raster or something. Another thing to watch out for is that the GDAL version provides somewhat lower quality rasters when displaying zoomed-out versions. I don't really know whether this is a DotSpatial problem or a GDAL problem. I will have to look a bit further into this.

@TongZhai you could give all this a shot and let us know your results!

**Comment by jany-tenaj** [comment]
@TongZhai The thought of having DotSpatial do everything itself is nice as long as you know how to fix the errors that might come up. If you're not able to fix them its much easier to use things like gdal functions that get fixed by someone else. ;) That goes for a lot of other code, too. Basically we had DotSpatial.Topology that was a copy of NTS and we have Projections that is a port of Proj.4 and so on. But we have no one that has the time to maintain all those projects to make sure they work correctly.

**Comment by TongZhai** [comment]
I tried the regular raster loading in DemoMap for the 2GB geotiff, the whole process took 7 minutes to generate the mwi and mwh files, after these files are created, when opening the same grid again, it took 21 seconds. So, generating those two files took most of the time. There are errors in reading the data as shown below (see those weird band):

![image](https://user-images.githubusercontent.com/16242714/80852761-0e39ad80-8bf9-11ea-8e3c-80f77540af83.png)

I also tried open the same grid using the GdalImageProvider.Open, it displayed the below image immediately

![image](https://user-images.githubusercontent.com/16242714/80852788-53f67600-8bf9-11ea-8148-2f89b54a23fa.png)

I checked, it didn't generate the .mwi and .mwh files, and when I zoom on the loaded grid, it failed as below:

![image](https://user-images.githubusercontent.com/16242714/80852812-830ce780-8bf9-11ea-8bf8-8aa0cc631aff.png)

**Comment by TongZhai** [comment]
@jany-tenaj I didn't realize until now that DotSpatial has its own Raster/pyramid scheme. In this instance, if the `DotSpatial.Data.Rasters.GdalExtension` is truly a wrapper of the Gdal libraries, and assuming DotSpatial is using Gdal libraries the correct way, then it is Gdal's issue perhaps, and other GIS such as QGis and MapWindow could be doing something different than Gdal. I understand that it takes time and effort to make things better in DotSpatial, and everybody is busy, so let's start small, one step at a time, can we at least get @VectorZita's issue fixes merged into main branch, also let's fix that file pointer math problem, which we already verified that they made things better. Beyond that, we could start a list of development tasks and provide necessary code documentation such that other developers can contribute. @VectorZita I could try that skipping zero-scale pyramid approach, not sure where in the code to change. I personally might have some time available later half of this month, so if you need help working on code changes, we can collaborate.

**Comment by jany-tenaj** [comment]
@TongZhai We might have to update to the latest gdal libraries, too.

@TongZhai @VectorZita Have you made sure that the fix Vector provided doesn't cause any side effects somewhere else?

**Comment by VectorZita** [comment]
@jany-tenaj In #1326 I suggested a correction in the `DotSpatial.Data.EnvelopeExt` class `Reproportion` method. As far as I have seen, this method is only called from the `PyramidImage` class, in the `public override GetBitmap(Extent, Rectangle)` method. Similarly, the `DotSpatial.Symbology.ExtentExt` class `Reproportion` method is only called in `MapImageLayer` class `DrawWindows(...)` method.

In #1332 I suggested changing one method and introducing a new method, which performs something slightly different (i.e. returns negative values for cell row/column indices). The method I suggest a change of is the `DotSpatial.Data.RasterBoundsExt` classs `CellsContainingExtent` method. This is only used in `PyramidImage` class `GetBitmap` method as well.

Therefore, the only side effects of my proposed changes currently affect `PyramidImage.GetBitmap(Extent, Envelope)` and `MapImageLayer.DrawWindows(MapArgs, IList<Extent>, IList<Rectangle>` methods.

The only validation I have at the moment (beyond my own testing, which is why and how I came up with those fixes) is that @TongZhai has had a problem using the `PyramidImage` class before applying the fixes, which was gone after applying them. In the meantime, I don't know if the same situation was checked against the `MapImageLayer`. @TongZhai as I said above, the change in `DotSpatial.Symbology.ExtentExt` affects the `MapImageLayer`. Have you tried opening with the MapImageLayer before and after applying the change? As far as I can see, things are slightly more complicated there and this is the only thing the behavior of which I have not analyzed extensively (and therefore not certain about).

I will drop a test over there soon as well. @TongZhai mind that the old GdalImageProvider commit that utilizes GDAL to build the pyramids **actually changes your raster file**.

@TongZhai regarding your problem in `GdalImage`, can you please reproduce it and check the values of `xSize` and `ySize` that were passed in the `LockBits` method? These are almost definitely the source of the problem. There is a call a few lines earlier:

`NormalizeSizeToBand(xOffset, yOffset, xSize, ySize, firstO, out width, out height);`

This has probably produced `width` and `height` outputs that are less than the requested `xSize` and `ySize` rectangle to lock in the `LockBits` call, thus the _traditional_ GDI+ error ("parameter is not valid").

This is, of course, just by skimming through the code and is, likely, only the mechanism. I will have to look into this a bit more.

@jany-tenaj as far as I can see, `GdalImage` lacks some error-checking and prevention. In my experience, `Marshal`-ing and using external bindings can produce all sorts of errors at almost any time. I usually tend to try to get away with an occasional drawing artifact rather than having an application crash. Do you think some `try-catch` blocks and error checks (e.g. checking the returned `CPLErr` of some GDAL calls) should be added in some places to avoid crashes at the expense of drawing artifacts?

**Comment by jany-tenaj** [comment]
@VectorZita I'm all for catching errors instead of popping them up to the upmost level. It's annoying for end users to see those errors. And who is going to fix them up there?

The question is why could those functions create artefacts? Could we repair those images? If not, I'd at least suggest to inform the user that the function didn't work as expected or scrap the whole thing because it didn't work.

**Comment by TongZhai** [comment]
@VectorZita This is strange, I just tried opening the large grid with `GdalImageProvider.Open` again, and it didn't die in that `GdalImage` file anymore. It actually loaded the grid in the same way as the regular grid open (with the weird bands and all). This is when the .mwi and .mwh already exist. When I removed those two files, it actually died with the same file pointer exception as shown below:

![image](https://user-images.githubusercontent.com/16242714/80872196-6797f000-8c7e-11ea-88ff-84fcb969e851.png)

What's more, it didn't seem to internally modify the grid file as its size didnot increase and it is attempting to generate the .mwi and .mwh files.

**Comment by TongZhai** [comment]
@VectorZita the values of `xSize` and `ySize` that were passed in the `LockBits` method are 257 and 220 respectively, at the time of failure.

The line: `NormalizeSizeToBand(xOffset, yOffset, xSize, ySize, firstO, out width, out height);` has the following argument values:
xOffset: 871
yOffset: 499
xSize: 257
ySize: 220

The firstO object has a few exceptions as below:

![image](https://user-images.githubusercontent.com/16242714/80873475-c2cde080-8c86-11ea-8e6c-d59a307be74e.png)
